(PUP-9994) implement $trusted['external'] facts

DavidS opened the pull request:
These facts are supplied through a new external lookup. They can be used
for injecting additional facts from third-party databases like asset
management or service-now.

To configure a lookup, provide the name of your lookup command to
`trusted_external_command`. The lookup script will receive the certname
as first and only argument and is expected to return the external facts
as a JSON hash on stdout.

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on lib/puppet/indirector/trusted_ext/plain.rb:
```diff
@@ -0,0 +1,11 @@
+require 'puppet/indirector/plain'
+
+class Puppet::TrustedExt::Plain < Puppet::Indirector::Plain
+  desc "Always return an empty Hash. This is a placeholder to make the default configuration a noop."
```
Can you default the terminus to `:none` instead? There's already a `Puppet::Indirector::None` terminus whose find method returns nil.

DavidS replied:
done.

justinstoller commented:
FWIW, I know of a customer who wanted functionality like this. They originally tried to implement it as a `Compiler` terminus (but that breaks `CD4PE`); they tried it as a `Node` terminus but when eliminating the round tripping of facts to disk before a compile we stopped calling `#find` on the `Node` terminus everywhere and now in some paths we only deserialize the node's facts and then directly act on those values. I believe the customer is now monkeypatching fact deserialization, which is far from ideal.

I don't know if this is a good implementation to provide that functionality (I'll leave that question to y'all), but I know of some business use cases that support _something_ to this effect.

At a higher level the customer workflow requires something like group data in the console but at a per-node level that is more scalable than anything else we can do right now.

Regardless, of the implementation details of this, they will have repercussions for Puppet Server's CD4PE, ACE, and apply-in-plans endpoints which do custom compilations and don't go through all of the same codepaths as agent compilations.

mrzarquon commented:
This was an experiment from talking with @DavidS - it is trying to codify the ability to add trusted external data for use by nodes, similar to the trusted data today, but instead of coming from a certificate, it would be loaded from another source (hence external) at agent check in, and fed in as facts. This got done before the associate ticket could be put up.

DavidS commented:
> Why is this feature needed? 

Customer demand. Please talk to @mrzarquon for details.

> Couldn't you create a node terminus that composes the normal node terminus and this new thing?

How would that make the UX or implementation better?

> If we add a new way of retrieving trusted data, but merge those values into facts, then you have to be careful when creating a Node object using cached fact values. See the Node#sanitize method.

There is already a change in the `sanitize` method. Is that sufficient or do you think more is needed there?

> Regardless, of the implementation details of this, they will have repercussions for Puppet Server's CD4PE, ACE, and apply-in-plans endpoints which do custom compilations and don't go through all of the same codepaths as agent compilations.

Can you point me to how these endpoints are doing compilation? I covered all code-paths I could find in the core code-base.

> Please add the ticket name to the title of the PR

done.

DavidS commented:
Updated based on the internal conversation.

DavidS commented:
Added `--trace` debugging to the exec failure on windows, but can't make head or tails of it. maybe someone from @puppetlabs/windows can help?

DavidS commented:
Given that the same thing failed on JRuby with `fork is not available on this platform`, it might not be a windows problem after all. I'll continue digging.

DavidS commented:
As far as I can tell from https://github.com/puppetlabs/puppetserver/blob/074fe9701ceff3e62cb17889ec768840214d143b/src/ruby/puppetserver-lib/puppet/server/execution.rb `Puppet::Util::Execution.execute` is overridden on JRuby by puppetserver, henceforth anything using it is not testable on JRuby in the main puppet repo. Still don't know what's broken on Windows, but all of that is now a question for Monday. 

:wave:

mrzarquon commented:
For historical purposes: the change from $trusted_ext to $trusted["external"] was to keep changes minimal in the rest of the Puppet tool chain. We already know how to handled the $trusted hash, so instead of adding a new one, we're adding a new top level key "external" to it. This was the result of a conversation with @joshcooper, @hlindberg and myself.

glennsarti commented on spec/integration/indirector/trusted_external/exec_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  context "with a trusted_external_command configured" do
+    let(:script_path) do
+      script_containing('script',
+        windows: "@echo '{ \"fact\": \"ABCD\" }'"
```
Single quotes are *NOT* a string delimiter in `cmd.exe`

In fact, echo will literally spit out the text anyway

```
C:\Users\glenn.sarti>@echo { "fact": "ABCD" }
{ "fact": "ABCD" }

C:\Users\glenn.sarti>@echo '{ "fact": "ABCD" }'
'{ "fact": "ABCD" }'
```

glennsarti commented on lib/puppet/indirector/trusted_external/exec.rb:
```diff
@@ -0,0 +1,19 @@
+  def command
+    command = Puppet[:trusted_external_command]
+    raise ArgumentError, _("You must set the 'trusted_external_command' parameter to use the extended trusted facts terminus") unless command != "none"
+    command.split
```
If this is supposed to run on Windows, `command.split` is probably a very naive way to split a string into an array of strings to execute. For example; How would you pass `C:\Program Files\ServiceNow\fetchnode.exe` ?

glennsarti commented on lib/puppet/defaults.rb:
```diff
@@ -536,6 +536,33 @@ def self.default_vendormoduledir
+        * `none` --- Returns no data.
+        * `exec` --- Uses an external trusted facts script configured
+          by the `trusted_external_command` setting. This lets you pull a
+          list of external trusted facts, using a small glue script to
```
nit: It's not a `script` it's an `executable command` as per the docs below.

joshcooper commented on lib/puppet/trusted_external.rb:
```diff
@@ -0,0 +1,9 @@
+  # Set up indirection, so that external trusted facts can be loaded
+  extend Puppet::Indirector
+
+  indirects :trusted_external, :terminus_setting => :trusted_external_terminus, :doc => "Where to find node external trusted facts."
```
I'd prefer to not add new indirections. Can we just add a `trusted_command` setting of type `:file`, which defaults to `nil`, similar to the `autosign_command`?

DavidS replied:
This would shut the door on (easily) adding a REST indirector further down the line. @mrzarquon would that reduction in flexibility still work for your use-case?

mrzarquon replied:
I'm ok with focusing just on getting the exec working and using future httpclient feature if we discover we need a faster way to pull data from a rest interface.

joshcooper commented on lib/puppet/context/trusted_information.rb:
```diff
@@ -51,17 +61,42 @@ def self.remote(authenticated, node_name, certificate)
+    else
+      raise Puppet::Error, _("Unsupported data type: '%{klass}'") % { klass: object.class }
+    end
+    object
```
I see this is copied from the parser scope. But puppet vendors `deep_merge`! I'm ok adding this duplication for now. I filed https://tickets.puppetlabs.com/browse/PUP-10014.

DavidS replied:
I concede the point on code duplication - and I'd be up for pulling out the function to `Util` if this goes forward. Re Pup-10014, this is `deep_freeze`, not `deep_merge`.

joshcooper replied:
wow, total :brain fart: on my part

DavidS commented:
@glennsarti do you think that any of your remarks will fix the issue on appveyor? I'm currently at a conference, so can't look into details today.

glennsarti commented:
@DavidS I suspect `windows: "@echo '{ \"fact\": \"ABCD\" }'"` is causing failures because that is not a valid JSON string

ody commented:
How does this change the meaning of trusted facts? Once a fact is mutable and especially by an outside system, how much of a promise can be provided that the fact is truly trusted?

Are these better defined as compiler sourced/retrieved external facts? Essentially a way to obtain pieces of metadata about a node from a CMDB that the node doesn't have direct access to.

reidmv commented:
I don't think this changes the semantics of "trusted". Trusted just means the node can't change it. "Trusted External" even in plain language implies trust of whatever "external" is. Misplacing trust is totally possible, but the core semantic of "the node can't change this" isn't broken.

ody commented:
The node can't change it in the context of Puppet but external sources of data are often updated dynamically with data pushed from an agent, e.g. Consul. I could also argue that ServiceNow Discovery which can populate the ServiceNow CMDB, while agentless is still affected by node generated pieces of data.

DavidS commented:
* removed indirector and replaced by a simple command
* rebased on current master
* still needs some kind of integration/acceptance test, if this approach is OK

joshcooper commented on lib/puppet/trusted_external.rb:
```diff
@@ -0,0 +1,13 @@
+module Puppet::TrustedExternal
+  def retrieve(certname)
+    command = Puppet[:trusted_external_command]
+    return nil if command == "none"
```
Can the setting default to `nil` so we don't have to check for `none`?

DavidS replied:
fixed

TJM commented:
We could really use functionality like this as well. The ability to pull service now data in for reference during a puppet run seems very powerful. We have been contemplating several different ways of accomplishing this, but this one seems like it has promise. I fear that using `certname` as the one and only argument may be a bit limiting. Just because the default value for `certname` is the FQDN of the host doesn't necessarily mean it will always be FQDN, nor is it always ideal in all cases (think cloud deployments). I have seen a GUID used for certname in places where hostnames are reused or unreliable. It might be nice if the argument was "configurable" as any portion of the $trusted hash or perhaps with the ability to pass any (or all) of the "$trusted" hash to STDIN?

davidmalloncares commented:
@joshcooper - would you have time to review and merge this in if it looks ok from your perspective? We have tested here in the Compliance team and it works beautifully for us and we would really need it merged in to unblock us for the next stage of our development

davidmalloncares commented:
@mrzarquon - do you need to close your review of this PR out?

joshcooper commented on lib/puppet/defaults.rb:
```diff
@@ -306,7 +306,7 @@ def self.default_vendormoduledir
         :type     => :boolean,
         :desc     => "Whether Puppet should manage the owner, group, and mode of files it uses internally.
-        
+
```
nit unrelated whitespace

joshcooper commented:
@DavidS I'm fine with the implementation (left a small nit to fix up). Can you add a test for this? There is an autosign beaker test that you could probably borrow from. /cc @davidmalloncares

joshcooper commented:
@TJM Agree that the certname isn't always the client's FQDN. It can be a UUID, but puppet doesn't really care. All that matters is that it's the node's identity, which allows us to write puppetserver ACLs based on the identity, generate logs, etc. In this PR, I would expect the node's identity (aka certname) to be passed to the external script. Is there a situation where the script would need something other than the identity (or in addition to)?

davidmalloncares commented:
Thanks @joshcooper , @tphoney is taking over the PR and will hopefully get a look at it this week

reidmv commented:
The prior art we have in the Puppet tool for this kind of interface, in certificate autosigning, is to invoke the user script with a single positional argument being the node's identity, and if any additional information is required to pass that along on stdin.

I suggest we don't deviate from that pattern. I further suggest that we don't bother implementing the passing of additional information on stdin right now. Let's ship version 1 as-is, passing a single positional argument to the user script (the node identity).

We can reserve the option of passing additional info on stdin later, if compelling use cases are identified. IMO, it's a bit of a stretch to try and imagine such use cases at the moment.

TJM commented:
Agreed, lets release and iterate if needed.

TJM commented:
> @TJM Agree that the certname isn't always the client's FQDN. It can be a UUID, but puppet doesn't really care. All that matters is that it's the node's identity, which allows us to write puppetserver ACLs based on the identity, generate logs, etc. In this PR, I would expect the node's identity (aka certname) to be passed to the external script. Is there a situation where the script would need something other than the identity (or in addition to)?

Sorry, I did not see this last week. I guess I was trying to decide how I would tie a "cloud" based system where hostnames and IP addresses were more ephemeral to a "legacy" system like service now, where it seems they are still thinking in a bare-metal world :) ... there would have to be some way to look up the host in the target system by the identity argument that was passed. I guess we can leave that up to the "end-user" (script developer) as they should know the target system.

kris-bosland commented:
Superseded by #7813
